azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -291,12 +289,150 @@ def test_already_connected(self, mocker, stage, op):
+
+    @pytest.mark.it(
+        "Sends down a new ConnectOperation if the op fails and the transport is no longer connected"
+    )
+    def test_failure_and_disconnect_new_connect_operation(
+        self, mocker, stage, op, arbitrary_exception
+    ):
+        stage.pipeline_root.connected = True
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 2
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+    @pytest.mark.it(
+        "Runs the operation a second time if the stage needs to reconnect after the op fails"
+    )
+    def test_failure_and_disconnect_resubmit_operation(
+        self, mocker, stage, op, arbitrary_exception
+    ):
+        stage.pipeline_root.connected = True
+
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 2
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+        # complete the connect op
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+        # make sure our op got sent down again
+        assert stage.send_op_down.call_count == 3
+        assert stage.send_op_down.call_args[0][0] is op
+
+    @pytest.mark.it("Fails the op if the op fails and the transport is still connected")
+    def test_failure_still_connected(self, mocker, stage, op, arbitrary_exception):
+        callback = op.callback_stack[0]
+        stage.pipeline_root.connected = True
+
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # fail
+        op.complete(arbitrary_exception)
+
+        # assert that the op failure went back to the caller
+        assert op.completed
+        assert op.error == arbitrary_exception
+        assert callback.call_args == mocker.call(op=op, error=arbitrary_exception)
+
+
+@pytest.mark.describe(
+    "AutoConnectStage - .run_op() -- Called with an Operation that requires an active connection (pipeline not connected)"
+)
+class TestAutoConnectStageRunOpWithOpThatRequiresConnectionNotConnected(
```

Comment:
>TestAutoConnectStageRunOpWithOpThatRequiresConnectionNotConnected [](start = 6, length = 65)

Rename to similar previous class name `TestAutoConnectStageRunOpWithOpThatRequiresConnectionPipelineNotConnected`